say: introduce internal on_log_level routines

Korablev77 opened the pull request:
Subj. Required for [flight recorder issue](https://github.com/tarantool/tarantool-ee/issues/201)

locker commented on src/lib/core/say.c:
```diff
@@ -1259,7 +1259,7 @@ log_vsay(struct log *log, int level, const char *filename, int line,
 		va_end(ap_copy);
 	}
-	if (level > log->level) {
+	if (!say_log_level_is_enabled(level)) {
```
Before this patch, we checked the level of the `log` object while now we check the global log level. This looks incorrect, because `log_vsay` deals with a `log` object. Why do you want to change this?

locker commented on src/lib/core/say.h:
```diff
@@ -326,8 +326,7 @@ CFORMAT(printf, 5, 6) extern sayfunc_t _say;
  */
 #define say_file_line(level, file, line, error, format, ...) ({ \
-	if (say_log_level_is_enabled(level)) \
-		_say(level, file, line, error, format, ##__VA_ARGS__); })
```
The log level check was here for a reason, I think: we don't want to evaluate the format if the message isn't going to come through. Evaluating all debug-level messages unconditionally may be expensive.

Korablev77 commented:
@locker I've reworked patchset in the way @alyapunov suggested (i.e. introduced new on_log_level var and left all checks so that there should not be any performance degradation).

locker commented on src/lib/core/say.h:
```diff
@@ -260,6 +260,10 @@ log_set_format(struct log *log, log_format_func_t format_func);
 
+/** Set log level for the on_log callback. */
+void
+say_set_on_log_level(int new_level);
```
Since this is needed only if there's a callback, better pass it as an extra argument to `say_set_log_callback`.

Korablev77 replied:
Ok, refactored.

locker commented on src/main.cc:
```diff
@@ -481,6 +481,7 @@ load_cfg(void)
 			       cfg_geti("flightrec_requests_size"),
 			       cfg_geti("flightrec_requests_max_req_size"),
 			       cfg_geti("flightrec_requests_max_res_size"));
+		say_set_on_log_level(flightrec_log_level());
```
This had better be done from `flightrec_init()` - so as not to introduce yet another stub.

Korablev77 replied:
Alongside with comment above - fixed.

locker commented on src/lib/core/say.h:
```diff
@@ -270,10 +270,10 @@ void
  */
 void
-say_set_log_callback(log_callback_t callback);
+say_set_log_callback(log_callback_t callback, int on_log_new_level);
```
Nit: the argument could be named simply `level`.

locker reviewed (approved):
Please don't merge until the EE part is ready to go.
(So as not to block EE development.)

locker commented:
Cherry-picked to 2.10.
